VAULT-29412: Preview of new Vault Radar resource.

trentdibacco opened the pull request:
__Preview__ of Manage a Vault Radar source of type GitHub Enterprise via the HCP Terraform Provider.

> [!IMPORTANT]  
> This is a preview, and my intent is for this resource not to show up in the in the [online documentations](https://registry.terraform.io/providers/hashicorp/hcp/latest/docs).
> Therefore, I did not set the **subcategory** in the markup.

## New Resource Checklist
- [x] __Plugin Framework__ 
- [x] __Minimal LOC__: **1 resource at a time**.
- [x] __Acceptance Tests__: 
- [x] __Documentation__: 
- [x] __Well-formed Code__: `go fmt`. 

### Schema
- [x] __Uses Globally Unique ID__: Uses GUID returned by server.
- [x] __Validates Fields Where Possible__: Minimal
- [x] __Supports Optional Project ID Input__: 

### CRUD Operations
- [x] __Uses Context-Aware CRUD Functions__: 
- [x] __Uses Context For API Calls__: 
- [x] __Uses DRY API Calls__: added vault_radar under client dir for reusable client ops.
- [ ] __Handles Existing Resource Prior To Create__: ? Server side will return an error during creation, no public api yet for checking prior to creating.
- [ ] __Implements Immediate Resource ID Set During Create__:  ? before other API operations or returning? Its using the GUID returned by the server side.
- [x] __Refreshes Attributes During Read__: most of these attributes are immutable.
- [x] __Handles Removed Resource During Read__: 
- [x] __Handles Failed State During Read__:
- [x] __Handles Nonexistent Resource Prior To Delete__:

### Documentation
- [x] __Includes Descriptions For Resource And Fields__: 
- [x] __Includes Example__: 
- [x] __Includes Generated Docs__: 

### :hammer_and_wrench: Description

Manage a Vault Radar source of type GitHub Enterprise via the HCP Terraform Provider.

### :building_construction: Acceptance tests

> [!WARNING] 
> This AC test is being skipped because of pre-existing conditions that would be require for the test to succeed. 
> - Requires Project already setup with Radar.
> - Requires a Service Account with an Admin role on the Project.
> - Requires access to a GitHub Enterprise Server self-managed instance.
> - Requires the following environment variables to be set:
>    - HCP_PROJECT_ID
>    - RADAR_GITHUB_ENTERPRISE_ORGANIZATION
>    - RADAR_GITHUB_ENTERPRISE_DOMAIN
>    - RADAR_GITHUB_ENTERPRISE_TOKEN

> [!NOTE]
> I've created a ticket to note how we can address this sometime in the future.
> https://hashicorp.atlassian.net/browse/VAULT-31051

- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch? 

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR. More info on acceptance tests here: https://github.com/hashicorp/terraform-provider-hcp/blob/main/contributing/writing-tests.md

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```sh
make testacc TESTARGS='-run=TestRadarSourceGitHubEnterprise'
==> Checking that code complies with gofmt requirements...
golangci-lint run --config ./golangci-config.yml 
TF_ACC=1 go test ./internal/... -v -run=TestRadarSourceGitHubEnterprise -timeout 360m -parallel=10
?   	github.com/hashicorp/terraform-provider-hcp/internal/clients/iampolicy	[no test files]
?   	github.com/hashicorp/terraform-provider-hcp/internal/clients/packerv2	[no test files]
?   	github.com/hashicorp/terraform-provider-hcp/internal/customdiags	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/clients	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/consul	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/hcpvalidator	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/input	(cached) [no tests to run]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/acctest	[no test files]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/customtypes	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider	(cached) [no tests to run]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/iam/helper	[no test files]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/modifiers	[no test files]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/packer	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider/iam	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider/logstreaming	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider/packer/datasources/artifact	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider/packer/datasources/version	(cached) [no tests to run]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/packer/testutils	[no test files]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/packer/testutils/configbuilder	[no test files]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/packer/testutils/configbuilder/packerconfig	[no test files]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/packer/testutils/testcheck	[no test files]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/packer/testutils/testclient	[no test files]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/packer/utils	[no test files]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/packer/utils/base	[no test files]
?   	github.com/hashicorp/terraform-provider-hcp/internal/provider/packer/utils/location	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider/packer/resources/bucket	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider/resourcemanager	(cached) [no tests to run]
=== RUN   TestRadarSourceGitHubEnterprise
--- PASS: TestRadarSourceGitHubEnterprise (17.23s)
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider/vaultradar	17.868s
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider/vaultsecrets	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider/waypoint	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider/webhook	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider/webhook/validator	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/providersdkv2	(cached) [no tests to run]

...
```

Commit:
VAULT-29412: Preview of new Vault Radar resource.

Manage a Vault Radar source of type GitHub Enterprise via the HCP Terraform Provider.

Commit:
Add to changelog.

Commit:
Skip automated acceptance tests.

trentdibacco commented on internal/provider/vaultradar/resource_radar_source_github_enterprise_test.go:
```diff
@@ -0,0 +1,50 @@
+	token := os.Getenv("RADAR_GITHUB_ENTERPRISE_TOKEN")
+
+	if projectID == "" || githubOrganization == "" || domainName == "" || token == "" {
+		t.Skip("HCP_PROJECT_ID, RADAR_GITHUB_ENTERPRISE_ORGANIZATION, RADAR_GITHUB_ENTERPRISE_DOMAIN, and RADAR_GITHUB_ENTERPRISE_TOKEN must be set for acceptance tests")
```
@jasonpilz Seems the automated testing is on, and this will fail for CI.
* Requires Project already setup with Radar.
* Requires a Service Account with an Admin role on the Project.
* Requires access to a GitHub Enterprise self-managed instance.

This works on my local setup with the proper env setup.

trentdibacco replied:
I created the following ticket to try an address this in the future.
https://hashicorp.atlassian.net/browse/VAULT-31051

jasonpilz replied:
Sounds good, thanks 👍🏻

trentdibacco commented on docs/resources/vault_radar_source_github_enterprise.md:
```diff
@@ -0,0 +1,45 @@
+---
+page_title: "hcp_vault_radar_source_github_enterprise Resource - terraform-provider-hcp"
+subcategory: ""
```
@jasonpilz I don't want the documentation to show up in the online docs, so I left `subcategory` blank.

trentdibacco commented on internal/provider/vaultradar/resource_radar_source_github_enterprise.go:
```diff
@@ -0,0 +1,235 @@
+		return
+	}
+
+	plan.ID = types.StringValue(res.GetPayload().ID)
```
@jasonpilz Note that I'm using the GUID generated from the server call, instead of generating something and setting the id first, which seem to be inferred by the new resources guide.

Is this okay?

kathytong-hashicorp commented on internal/provider/vaultradar/resource_radar_source_github_enterprise.go:
```diff
@@ -0,0 +1,235 @@
+				},
+			},
+			// Optional inputs
+			"project_id": schema.StringAttribute{
```
Should we validate UUID?

trentdibacco replied:
I don't think so, I didn't see anyone else doing that.

kathytong-hashicorp commented on docs/resources/vault_radar_source_github_enterprise.md:
```diff
@@ -0,0 +1,45 @@
+# hcp_vault_radar_source_github_enterprise (Resource)
+
+-> **Note:** HCP Vault Radar Terraform resources are in preview.
+
+This Vault Radar resource manages HCP Cloud scans of GitHub Enterprise sources.
+
```
nit: While we call it `github_enterprise` on the Radar side, can we specify somewhere that this is actually for GitHub Enterprise Server (I think the terminology is that "Enterprise" is just paid, so it could be GitHub Enterprise Cloud as well, even though we call that just `github_cloud`) 

The portal has added the differentiation as so as well

<img width="406" alt="Screenshot 2024-09-16 at 4 23 09 PM" src="https://github.com/user-attachments/assets/32e36fec-6f4b-4eb6-904e-29477bc7f5d1">

trentdibacco replied:
I can certainly update the docs, not sure if I want to change the resource name as its pretty long as is.

trentdibacco replied:
Update docs, didn't change the name of the resource.

kathytong-hashicorp reviewed (approved):
Can't say I understand much of the TF stuff but the Radar stuff looks good to me!

trentdibacco commented on internal/clients/vault_radar.go:
```diff
@@ -0,0 +1,115 @@
+		return err
+	}
+
+	deletionConformation := func() (bool, error) {
```
will correct.

Commit:
fix spelling error, minor tweak to docs.

kartheek-hc commented on internal/clients/vault_radar.go:
```diff
@@ -0,0 +1,115 @@
+func waitFor(ctx context.Context, retry, timeout time.Duration, maxConsecutiveErrors int, isDone func() (bool, error)) error {
+	consecutiveErrors := 0
+
+	waitCtx, cancel := context.WithTimeout(context.Background(), timeout)
```
Can we use the context from the function arguments?
```suggestion
	waitCtx, cancel := context.WithTimeout(ctx, timeout)
```

trentdibacco replied:
Good catch.

kartheek-hc commented on internal/provider/vaultradar/resource_radar_source_github_enterprise.go:
```diff
@@ -0,0 +1,235 @@
+
+func (r *sourceGitHubEnterpriseResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "This Vault Radar resource manages HCP Cloud scans of GitHub Enterprise Server sources.",
```
nit: Same here, managing scans is somewhat confusing. Consider changing to `This terraform resource manages GitHub Enterprise server datasource lifecycle in Vault Radar`?

trentdibacco replied:
The other spot is generated from here: As noted I will update this to: 

> This terraform resource manages GitHub Enterprise Server data source lifecycle in Vault Radar.

kartheek-hc commented on internal/provider/vaultradar/resource_radar_source_github_enterprise.go:
```diff
@@ -0,0 +1,235 @@
+					stringplanmodifier.RequiresReplace(),
+				},
+				Validators: []validator.String{
+					stringvalidator.RegexMatches(regexp.MustCompile(`^\S+$`),
```
nit: Can we have a more restrictive regex?

kartheek-hc commented on internal/provider/vaultradar/resource_radar_source_github_enterprise.go:
```diff
@@ -0,0 +1,235 @@
+				Description: "GitHub personal access token.",
+				Required:    true,
+				Sensitive:   true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
```
We don't need to destroy and recreate the datasource in Vault Radar to update the token right, is `RequiresReplace` the right modifier to use?

trentdibacco replied:
I don't have the update implemented yet.
I will need to expose a public api for that.
I will add that in a future PR.

kartheek-hc replied:
Sounds good, I didn't realize that update wasn't implemented when adding the comment

kartheek-hc commented on internal/provider/vaultradar/resource_radar_source_github_enterprise.go:
```diff
@@ -0,0 +1,235 @@
+type gitHubEnterpriseSource struct {
+	ID                 types.String `tfsdk:"id"`
+	DomainName         types.String `tfsdk:"domain_name"`
+	GitHubOrganization types.String `tfsdk:"github_organization"`
```
nit: Can we remove `GitHub` prefix since it is already a part of `gitHubEnterpriseSource`?

trentdibacco replied:
I don't want this to be confused with the HCP Organization.

kartheek-hc commented on internal/provider/vaultradar/resource_radar_source_github_enterprise.go:
```diff
@@ -0,0 +1,235 @@
+		return
+	}
+
+	// Resource is marked as deleted on the server.
+	if res.GetPayload().Deleted {
+		tflog.Info(ctx, "Radar source marked for deletion, removing from state.")
+		resp.State.RemoveResource(ctx)
+		return
+	}
```
I thought we want to remove the resource from state only if it is deleted. If we remove it while it is being deleted, could there be any side-effects? I am wondering about a case where the user runs `terraform destroy` and the deletion fails but we remove it from the state when they run `terraform plan` again

trentdibacco replied:
Make sense, I sync with you later but I think the best thing for this case then is not to update state, but maybe just log a warning.

kartheek-hc commented on internal/provider/vaultradar/resource_radar_source_github_enterprise.go:
```diff
@@ -0,0 +1,235 @@
+		return
+	}
+
+	// Resource is already marked as being deleted on the server.
+	if res.GetPayload().Deleted {
+		tflog.Info(ctx, "Radar source marked for deletion, removing from state.")
+		resp.State.RemoveResource(ctx)
+		return
+	}
```
Just curious, should we wait for the Vault Radar datasource to be deleted in this case?

trentdibacco replied:
Makes sense. I will refactor this to wait on the delete to complete if it was already initialize...

This also makes sense in the case that delete was called, inited, but didn't complete within the timeout of 10m, then they can run this again, picking up that its already in a delete in progress.

kartheek-hc reviewed (commented):
Looks good, have few some questions and suggestions

Commit:
Address some of Kartheek's PR feedback.

Update resource desc, fix context, better regex for validating domain name.

Commit:
Refine handling of Get and Delete of resource.

When the resource has been marked, but not fully deleted we need to handle
this case differently for Get and Delete.

Commit:
Minor fix to log message.

trentdibacco committed:
Merge branch 'main' into radar-VAULT-29412

jasonpilz reviewed (approved):
LGTM!

Commit:
update code owners.

Commit:
adj code owner.

Commit:
Merge remote-tracking branch 'origin/main' into radar-VAULT-29412

Commit:
Merge remote-tracking branch 'origin/main' into radar-VAULT-29412
